native/src/main/java/io/ballerina/lib/aws/mpe/AwsMpeClient.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com)
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.lib.aws.mpe;
+
+import io.ballerina.runtime.api.Environment;
+import io.ballerina.runtime.api.Future;
+import io.ballerina.runtime.api.values.BError;
+import io.ballerina.runtime.api.values.BMap;
+import io.ballerina.runtime.api.values.BObject;
+import io.ballerina.runtime.api.values.BString;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.services.marketplaceentitlement.MarketplaceEntitlementClient;
+import software.amazon.awssdk.services.marketplaceentitlement.model.GetEntitlementsRequest;
+import software.amazon.awssdk.services.marketplaceentitlement.model.GetEntitlementsResponse;
+
+import java.util.Objects;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * Representation of {@link software.amazon.awssdk.services.marketplaceentitlement.MarketplaceEntitlementClient} with
+ * utility methods to invoke as inter-op functions.
+ */
+public final class AwsMpeClient {
+    private static final String NATIVE_CLIENT = "nativeClient";
+    private static final ExecutorService EXECUTOR_SERVICE = Executors.newCachedThreadPool(new AwsMpeThreadFactory());
```

Comment:
> Better to add a `LinkedBlockingQueue` instead of `SynchronousQueue` (which is internally used in the `newCachedThreadPool()`) to increase the holding tasks before they are getting executed. Unless it will return an error to reject the newly added tasks when dealing with high loads.

Can you elaborate more on this ? Can't understand how a executor can reject a task when there are threads available.

Reply:
If there are sufficient threads available most likely a task won't get rejected. And according to this, having a `SynchronousQueue` will be fine if there is an unbounded `maximumPoolSize` to avoid task rejections
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadPoolExecutor.html#:~:text=Queuing,also%20decreases%20throughput.

But creating threads continuously can lead to resource exhaustion. IMO it would be better if we have  a thread pool with a env variable to manually configure the max thread pool size
https://github.com/ballerina-platform/module-ballerina-sql/blob/master/native/src/main/java/io/ballerina/stdlib/sql/datasource/SQLWorkerThreadPool.java
(Unlike here, we may need to use `LinkedBlockingQueue` for the workQueue)

Reply:
As per the cached thread pool doc [1] it can create new threads, re-use the idle threads, and kill the unused threads as needed. Hence, we don't need to manually handle this in our connector.

[1] - https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Executors.html#newCachedThreadPool--